attempt 1

RuiQianSun opened the pull request:
## What changes are you trying to make? (e.g. Adding or removing code, refactoring existing code, adding reports)
I am trying to complete assignment 1

## What did you learn from the changes you have made?
I learned how to find the first duplicate in a tree using level order traversal

## Was there another approach you were thinking about making? If so, what approach(es) were you thinking of?
Yes, I thought about converting the tree into a list first

## Were there any challenges? If so, what issue(s) did you face? How did you overcome it?
No challenges

## How were these changes tested?
I tested the code in VS Code

## A reference to a related issue in your repository (if applicable)
N/A

## Checklist
- [X] I can confirm that my changes are working as intended

RuiQianSun committed:
attempt 1

tianyi21 reviewed (changes requested):
Hi Rui Qian! 

Thank you for completing and submitting your ADS A1:
* You implemented an algorihm based on `BFS` to identify the duplicate element in a given tree.
* You utilized the `queue` and `dict` data structures in your solution.
* You demonstrated how you approached the problem and analyzed your solution.

I noticed the following issue(s) with your submission. I would appreciate it if you could **provide a short explanation** and **revise your submission** to address these issue(s):
* In your own example, when the input is `[1,2,2,1,1,1]`, there are multiple duplicates in this tree. In this case, the question asks for the closest one to the root, but your solution gives `2` in this case.
* In your alternative solution, you stated that this could also be done by converting the tree into a list and counting the occurrences. This is similar to a `DFS`-based algorithm. In this case, when there are multiple duplicates, is there a way to differentiate them?

Thanks,
Tianyi [LS]

RuiQianSun committed:
attempt 2

RuiQianSun commented:
Hello, I've updated my code.

> * In your own example, when the input is `[1,2,2,1,1,1]`, there are multiple duplicates in this tree. In this case, the question asks for the closest one to the root, but your solution gives `2` in this case.
_Understood, I've corrected the code. Given the input of BFS `[1,2,2,1,1,1]` the code now correctly returns `1`._

> * In your alternative solution, you stated that this could also be done by converting the tree into a list and counting the occurrences. This is similar to a `DFS`-based algorithm. In this case, when there are multiple duplicates, is there a way to differentiate them?
_I've updated my alternative solution to a DFS-based one which I think is better. In regards to my previous alternative solution I was thinking of using BFS to append nodes to a list so that the sorted order of the list would represent proximity to the root and then performing operations on that list to produce the closest duplicate._

tianyi21 reviewed (changes requested):
Hi Rui Qian,

Thank you for revising your assignment.

Unfortunately, the issue still exists in your alternative solution with the `DFS` algorithm. When you implement this with a stack and `DFS`, will the first one guarantee to be the closest one? Consider the following case `root=[1,2,3,4,5,3,7,4]`. What will be the **first** duplicate and what is the **closest**?

Could you please comment below how you would do to modify your current version so that it will always return the closest one? You can just briefly describe your strategy below and you don't have to resubmit your assignment.

Thanks,
Tianyi [LS]

RuiQianSun commented:
> Unfortunately, the issue still exists in your alternative solution with the `DFS` algorithm. When you implement this with a stack and `DFS`, will the first one guarantee to be the closest one? Consider the following case `root=[1,2,3,4,5,3,7,4]`. What will be the **first** duplicate and what is the **closest**?

> Could you please comment below how you would do to modify your current version so that it will always return the closest one? You can just briefly describe your strategy below and you don't have to resubmit your assignment.


Hello,

Ah I think I understand, thanks for pointing this out. In the current alternative DFS solution I actually would not know which duplicate is the closest. I believe the correct strategy is to use DFS (Preorder Traversal) to construct a dictionary to log the depth of each node value when the node is popped from the stack. Then I can perform some operation on that dictionary to return the closest duplicate to the root. I was able to get this to work last night, details below.

1. Create a function called "is_duplicate"
2. Create an empty dictionary called "seenDict" which takes {node value (int): depth (int)}
3. Create an variable called "result" to store the "closest duplicate" (int), assign -1 to "result"
4. Create a variable called "minDepth" (int) and set the value to positive infinity, this will help determine the "closest" duplicate
5. Create a nested function called "dfs" that takes "node" (TreeNode class) and "depth" (int)
6. Call dfs(root, 0)
7. Check if the value of the node is in "seenDict", if not, add the value and the depth to "seenDict"
8. call dfs(node.left, depth + 1)
9. call dfs(node.right, depth + 1)
10. if the value of the node is already in "seenDict", compare the depth of the node already in "seenDict" and the depth of the node that is currently being popped from the stack and take the "minimum depth of both occurrences"
11. If the "minimum of depth of both occurrences" is less than "minDepth", change minDepth to be the "minimum depth of both occurrences" and reassign "result" to the node value that is currently being popped
12. This solution will go through every node in the tree and update "result" every time it finds a duplicate with a lower depth, hence returning the "closest" duplicate

tianyi21 reviewed (approved):
Hi Rui Qian,

Thank you for the update! Your A1 is now complete.

Thanks,
Tianyi [LS]
